### PR TITLE
Compile compiler-rt with MSVC

### PR DESCRIPTION
**Pull request by retep998**
Compiles just the minimal subset of compiler-rt needed for MSVC and it does so using MSVC.

Depends on https://github.com/rust-lang/compiler-rt/pull/10

### CODE REVIEW

**Comment by huonw** [comment]
I suspect this will either optimise less well or be less precise (or both), so maybe it makes sense to have a `cfg` to only use this form on MSVC?

**Comment by retep998** [comment]
The compiler-rt implementation is pretty similar, so it should optimize just as well unless the platform has a native instruction for it. I can cfg it out to only be used on msvc though.

File: src/libcore/num/f32.rs [hunk 1]
```diff
@@ -257,7 +257,14 @@ impl Float for f32 {
 
     #[inline]
     fn powi(self, n: i32) -> f32 {
-        unsafe { intrinsics::powif32(self, n) }
+        let (mut r, mut a) = (1., self);
+        let mut b = if n < 0 { n.wrapping_neg() as u32 } else { n as u32 };
+        loop {
```
Comment (huonw) [inline src/libcore/num/f32.rs]: I would personally prefer this as the following, which is simpler (IMO):

```rust
while b != 0 {
     if b % 2 != 0 { r *= a }
     b /= 2;
     a *= a;
}
return if n < 0 { r.recip() } else { r }
```

But I don't feel particularly strongly. (Yes, I know the code as written does an additional `a *= a`.)

If you feel strongly about avoiding that extra `a *= a` I even prefer this form (i.e. basically just avoid folding the `return` into a complicated expression in the loop):

```rust
loop {
     if b % 2 != 0 { r *= a }
     b /= 2;
     if b == 0 { break }
     a *= a;
}
return if n < 0 { r.recip() } else { r }
```

**Comment by huonw** [comment]
> it should optimize just as well unless the platform has a native instruction for it

I'm talking about optimising at a higher level than just codegen, e.g. constant folding/reasoning about variable ranges etc.

Also, if it's a similar implementation, that means my precision concern isn't a concern.

**Comment by retep998** [comment]
Okay, I switched to using `cfg` so that the pure Rust version is only used for msvc.

**Comment by angelsl** [comment]
Perhaps we should do something like [rlibc][1] and continue to emit the opcodes, but instead of linking to `compiler-rt/builtins` we provide our own.

`compiler-rt/builtins` provides the things listed [here][2] but we really only need to implement those that we emit/we allow LLVM to emit.

Of course the other option is to get `compiler-rt/builtins` to build on MSVC, but that is going to be hell because it uses parts of C99 that MSVC doesn't support.

[1]: https://github.com/rust-lang-deprecated/rlibc/blob/master/src/lib.rs
[2]: https://gcc.gnu.org/onlinedocs/gccint/Soft-float-library-routines.html#Soft-float-library-routines

**Comment by angelsl** [comment]
Yeah, this works on our current nightlies. You can verify that `compiler-rt` isn't being depended on by replacing it with an empty `compiler-rt.lib` file (i.e. `type NUL > empty.c`, `cl /c /Tcempty.c`, `lib empty.obj`).

````
pub fn main() {
    println!("{}", 2.0f32.powi(4));
}

#[no_mangle]
pub extern fn __powisf2(mut a: f32, mut b: i32) -> f32
{
    let recip = b < 0;
    let mut r = 1f32;
    loop {
        if b & 1 == 1 { r *= a; }
        b /= 2;
        if b == 0 { break; }
        a *= a;
    }
    if recip { 1f32/r } else { r }
}

#[no_mangle]
pub extern fn __powidf2(mut a: f64, mut b: i32) -> f64
{
    let recip = b < 0;
    let mut r = 1f64;
    loop {
        if b & 1 == 1 { r *= a; }
        b /= 2;
        if b == 0 { break; }
        a *= a;
    }
    if recip { 1f64/r } else { r }
}
````

**Comment by angelsl** [comment]
Is there a compiler flag to set no_compiler_rt? Because if we simply disable compiler_rt from being built immediately then build fails because our stage0 compiler expects compiler_rt to exist.

Otherwise the solution is to apply this for one nightly and then actually stop building it the next.

**Comment by retep998** [comment]
@angelsl We'd have to wait till the next snapshot, not nightly, to disable compiler-rt from building.

**Comment by angelsl** [comment]
> @angelsl We'd have to wait till the next snapshot, not nightly, to disable compiler-rt from building.

Bah. I'll just put in my hackfix then.

**Comment by retep998** [comment]
I just did a simple test of checking the emitted asm and constant folding seems entirely unaffected.
https://gist.github.com/retep998/969089222a8ac84646ac

**Comment by alexcrichton** [comment]
Unfortunately I don't think we can take this step just yet. We need a guaranteed way to inform LLVM that it *can't* lower anything to a call to compiler-rt, otherwise it's likely to pop up from time to time and we'll have to go patch up various implementations here and there as they arise.

For example, a 32-bit target requires the `__mulodi4` intrinsic for a function like:

```rust
fn foo(a: i64, b: i64) -> i64 {
    a.checked_mul(b).unwrap_or(0)
}
```

It looks like other commonly used intrinsics from compiler-rt, such as for operations like `u64 % u64` on 32-bit, are wired up to functions that are found in the CRT, so we're covered on some fronts, but not all.

I agree it'd be nice to not have to build compiler-rt on MSVC, but I don't think it's worth "just hoping" that LLVM doesn't start lowering to any functions in compiler-rt. Another example of that currently LLVM doesn't lower expressions like `A * A * A * A * A * A` to a call to the `powi` intrinsics, but [it would be nice to do so](https://github.com/llvm-mirror/llvm/blob/c848236c936decfe077c92bf6232cf0feb913b26/lib/Target/README.txt#L45-L55). In that case we'd have to tell LLVM that the intrinsic simply isn't available rather than just not calling it ourselves.

**Comment by retep998** [comment]
Perhaps instead of avoiding using the intrinsics, we should just write pure Rust implementations of the intrinsics themselves?

**Comment by alexcrichton** [comment]
> Perhaps instead of avoiding using the intrinsics, we should just write pure Rust implementations of the intrinsics themselves?

Unfortunately even that isn't enough to cut it unless we start linking as all the weird intrinsic names (which they themselves are subject to change). LLVM frequently lowers to new intrinsics or creates instructions later lowered to new intrinsics as the result of optimizations, so we need a way to tell LLVM that it cannot perform these optimizations or lowerings by letting it know the intrinsic functions are not available.

> Considering you can use Clang with MSVC without compiler-rt, it should be entirely feasible for Rust to do so as well.

This is good news to know! It should help in learning what they do so we may be able to mirror it. Taking pot shots at the compiler, however, saying "X supports Y, why doesn't Rust" may not be so productive.

**Comment by retep998** [comment]
With regards to `__mulodi4`, Clay was having similar issues https://github.com/jckarter/clay/issues/164. The equivalent msvc intrinsic is `_mulx_u64` which isn't defined for 32-bit targets, and since C/C++ has no way to do multiplication with overflow (aside from those intrinsics) neither Clang nor msvc have to deal with this, so there's not much in the way of prior art in this area.

I doubt LLVM will ever start doing optimizations that lower to intrinsics which msvc doesn't provide since Clang is supposed to work without compiler-rt in the msvc toolchain. Thus it should be enough for us to simply not emit intrinsics which don't have a native instruction or msvc function and just have our own implementation instead.

**Comment by retep998** [comment]
Apparently D faced a similar issue and falls back to implementing multiplication with overflow themselves. https://github.com/ldc-developers/druntime/blob/c5a1f3186ffe7e0bcbbcb291d9d0df9c7d668614/src/core/checkedint.d#L584-L600

**Comment by alexcrichton** [comment]
Interesting! I wonder if perhaps if the LLVM folks have much to say about this? LLVM seems to claim that generated code that it should work without compiler-rt, but that seems contradictory given the current situation of how some intrinsics are lowered. It may be worth asking them if there's a precise set of intrinsics that are expected to *not* work or if there's a set which are intended to *only* work, and perhaps we can work with that?

Inlining implementations where necessary seems fine for now, and we may also want to have them with `#[cfg(target_env = "msvc")]` to be safe, but I'm mostly just concerned about the set of intrinsics we need to guard against.

**Comment by angelsl** [comment]
I think LLVM assumes the intrinsics are there, there is a way to set some intrinsics names to `nullptr` which seems to stop them from being emitted (?) but it doesn't seem to apply to all intrinsics. I'll dig through LLVM more when I find time..

I believe if we simply do not emit any `llvm.*` intrinsics we won't get any calls to compiler-rt (only to e.g. memset etc which for us now are in MSVCRT)

Someone on LLVM IRC yesterday said that "in a freestanding environment there are a subset of intrinsics that the user is expected to provide" and that it's documented in GCC, .. but I couldn't find anything. He might have misunderstood. Seems like this part of LLVM is pretty unknown :/

**Comment by retep998** [comment]
I decided to instead take the route of just compiling the necessary bits of compiler-rt with MSVC. Only one small change was needed to compiler-rt due to not needing to compile most of it. The changes to the make stuff are rather hairy and I don't really know if they're correct or robust. I hate make and I hope it gets replaced with something much easier to understand.

**Comment by angelsl** [comment]
Turns out we can't actually tell LLVM to not emit calls to `__powi*`.

[Here](http://code.metager.de/source/xref/llvm/llvm/lib/CodeGen/TargetLoweringBase.cpp#89) you can see that some intrinsics that are not always available are simply set to `nullptr`. But only uses of _those_ intrinsics check that they are available [(e.g. here)](http://code.metager.de/source/xref/llvm/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#2224) i.e. not `nullptr` &mdash; trying to do this for those we are concerned about i.e. `__powi*` results in LLVM segfaulting.

**Comment by angelsl** [comment]
OK, I got compiler-rt builtins compiling on MSVC.

Trying to upstream it: http://reviews.llvm.org/D13384

**Comment by alexcrichton** [comment]
Holy cow, nice work @angelsl! @retep998 would you be ok holding out to see the resolution of whether that gets merged upstream? It sounds like that'd be the ideal case!

**Comment by angelsl** [comment]
Should be merged soon, once I resolve some outstanding issues: [link](http://reviews.llvm.org/differential/query/u3OZ24cqaiIl/).

**Comment by angelsl** [comment]
Reviewer is merging the changes. They should be in tomorrow.

**Comment by alexcrichton** [comment]
Nice, thanks @angelsl! @retep998 this'll probably now want to start using `cmake` to build compiler-rt (we're already doing so for LLVM on MSVC), so it's probably fine to start basing this off a compiler-rt branch with those patches merged.

**Comment by angelsl** [comment]
Ah, one thing about `cmake`..

The `cmake` compiler-rt build expects either an already-compiled LLVM with the path to `llvm-config` passed to `cmake`, otherwise it should be located in the LLVM source tree in `projects/` and configured and built with LLVM.

So if we're compiling `rustc` with a prebuilt LLVM there is no problem, we just give the path to `llvm-config`.

But if we have to compile LLVM also we have to either build LLVM and then call `cmake` on `compiler-rt` later on with the path to the built LLVM, or put the compiler-rt submodule into `llvm/projects/` so it is built with LLVM.

Someone told me git supports having a submodule within the directory of another submodule (with both submodules of the same parent project) but I haven't tried that out.

The other thing is having all this MSVC-specific code so that we use `cmake` to build LLVM on MSVC and `automake` otherwise, but ... seems like LLVM is shifting towards using `cmake` wherever possible (including on Linux, etc) &mdash; this is an entirely separate issue though.

**Comment by angelsl** [comment]
Update: I forgot about the intrinsics written in assembly. LLVM's assembler files are in AT&T syntax while MASM only takes Intel notation ... MSBuild silently ignored the assembler files (!) so I didn't account for them until LLVM's buildbots which build with Ninja instead of MSBuild failed.

Now I can either choose to port the assembly files to Intel syntax which GAS will accept, or use the C implementations for the functions. Problem with the former is that that would mean LLVM has two different syntaxes for assembly in its source tree which isn't very nice.

I'm looking into using the C implementations when building with MSVC. The functions that are affected aren't actually used by Rust at the moment which is why when I tried the builtins library I built with those functions missing it still worked.

Alternatively, someone from LLVM said that they're working towards lowering libcalls to the MSVCRT instead. In that case we can continue using MinGW to build builtins until that works.

**Comment by angelsl** [comment]
All the changes except actually enabling the build on MSVC have been committed into the LLVM tree.

You can test it by changing [this line](https://github.com/llvm-mirror/compiler-rt/blob/master/lib/builtins/CMakeLists.txt#L350) to simply `else ()`.

There are some builtins for which the code output from C is pretty dismal compared to the handcrafted assembly output. Rust does not use any of them, however.

**Comment by alexcrichton** [comment]
Awesome, thanks @angelsl! @retep998 do you want to update this PR to use cmake to build compiler-rt as well?

**Comment by angelsl** [comment]
> 00:23:40 <%Retep9981> angelsl: I don't want to touch make
> 00:23:45 <%Retep9981> angelsl: make is scary

---

Rust currently builds LLVM on Windows by invoking CMake in `configure` and then starting the build later on when we actually run `make`. We also happen to treat `compiler-rt` separately from LLVM.

`compiler-rt` can be built as either standalone or with LLVM. When it is built with LLVM it should be located under `projects/` in the LLVM folder, and CMake will configure it along with the rest of LLVM when CMake is invoked for LLVM. When it is built standalone it expects LLVM to already be compiled and it requires the path to `llvm-config` passed to it when CMake is invoked on it.

So if we do not move `compiler-rt` we cannot call CMake for it during `configure` because we don't have LLVM yet, and calling CMake during actual `make` seems quite terrible. The other way is to move `src/compiler-rt` into `src/llvm/projects`... do we have a Git submodule for that path (so we have two, one for `src/compiler-rt` and `src/llvm/projects/compiler-rt` &mdash; yes, apparently Git supports this) (this means we have to update both submodules, bad) or do we move the directory (creating problems when Git is used later on because now its submodule is missing) or do we create a symlink (which we cannot do from Windows because that needs administrative rights (yay Windows) nor can we do from MSYS2 because that just creates a special file that the MSYS2 layer translates into a symlink for MSYS2 programs but Windows programs (e.g. MSBuild) won't know how to deal with)?

Yeah, this sucks.

**Comment by alexcrichton** [comment]
I think it's fine to call cmake as part of the normal build process, this is rarely build and builds quickly so it's not really any skin of anyone's back.

**Comment by angelsl** [comment]
@alexcrichton Oh, OK then.

Will do it the next few days.

**Pull request closed without merging**